### ISSUE TITLE
Es6 for of loops are fairly slow

### ISSUE THREAD

**Issue by leafpetersen** [dart-lang/sdk#30661]
I see a 20% speedup on a protobuf benchmark from changing (in Dart) a loop of the form:

```dart
for(var value in json) ...
```

to

```dart
var len = json.length;
for(int i =0; i < len; i++) {
  var value = json[i];
```

The main difference seems to be that the former uses es6 `for of` iteration, which seems to be slow.

Can we do better code generation here and still get the right semantics?

**Comment by lrhn** [comment]
Can you correctly detect that `json` is a *built-in* `List`? If so, the translation should be trivial, and if not, then it would likely be (potentially) incorrect.

**Comment by jmesserly** [comment]
yup, what @lrhn said. We need to use the iterator to be semantically correct, because the SDK does not expose any sort of `Array` type that we could know statically (array type(s) do exist in private implementation libraries, but users cannot access them directly). We could do some kind of local inference but that's about it.

You could try desugaring for-of to Dart iterators+while loop. My guess is that'd be slower, but might be worth a shot.

**Comment by vsmenon** [comment]
@leafpetersen - which version of V8?

From [V8 blog](https://v8project.blogspot.com/2017/09/elements-kinds-in-v8.html):

> Nowadays, the performance of both for-of and forEach is on par with the old-fashioned for loop.

**Comment by leafpetersen** [comment]
This was run using whatever versions of v8/chrome are used internally, I can check again soon.

We could in principle duplicate loop bodies for native Array/non array, but that can get expensive code size wise.  We could guide this with an annotation.

We could also provide an annotation for user code to serve as a checked assertion that we have a native Array.

**Comment by jmesserly** [comment]
> We could in principle duplicate loop bodies for native Array/non array, but that can get expensive code size wise. We could guide this with an annotation.

Yup. Early versions of dart2js tried a similar strategy (creating separate optimized/unoptimized code paths with bailouts), the code size hit was found to be too much, unfortunately.

> We could also provide an annotation for user code to serve as a checked assertion that we have a native Array.

That might be worthwhile, tho it seems less useful than an `Array<E>` type that you can write in code. I suppose an annotation lets us experiment in DDC without having to make the type public in the SDK.

While we're brainstorming, here's another idea: we could change the specified behavior of `for (var e in list) { ... }` when `list` has static type `List<E>`. We could allow (or require) that to be interpreted as `for (int i = 0, n = list.length; i < n; i++) { var e = list[i]; ... }`. The Dart 1 spec didn't take advantage of static type info, but Dart 2 could, in theory.

For a correct `List<E>` implementation, and assuming no concurrent modification, that is equivalent to the current behavior (as mentioned in the [API docs](https://api.dartlang.org/stable/1.24.2/dart-core/List-class.html)). Most users probably expect the indexing for loop and the for-in loop are equivalent. There's a subtle difference in terms of concurrent modification checking, although to my knowledge, neither dart2js nor DDC implement that correctly for arrays (too much performance cost to do so).

I wonder if this desugaring would have any performance benefit for other Dart implementations, especially in cases where they can't figure out that they've got an Array (either because of runtime polymorphism, or the limits of global type inference).

Anyways, fun to think about.
